### PR TITLE
SONARJAVA-1314 - Refactoring of NullConstraint

### PR DESCRIPTION
**Pull request by DidierBesset**

### CODE REVIEW

File: java-squid/src/main/java/org/sonar/java/se/SymbolicValue.java [hunk 1]
```diff
@@ -82,18 +74,18 @@ public int hashCode() {
 
   @Override
   public String toString() {
-    return "SV#" + id;
+    return "SV_" + id;
```
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/se/SymbolicValue.java]: really ?:)

File: java-squid/src/test/files/se/UnclosedResourcesCheck.java [hunk 2]
```diff
@@ -240,4 +245,16 @@ public void correctHandlingOfJarFile(String fileName) {
       closeJar(jar);
     }
   }
+
+  public void getDirectivesFromFile(File aFile) {
```
Comment (benzonico) [inline java-squid/src/test/files/se/UnclosedResourcesCheck.java]: This sounds unconnected to your refactoring and should not be part of this commit.

File: java-squid/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java [hunk 3]
```diff
@@ -140,12 +117,14 @@ public PostStatementVisitor(CheckerContext context) {
     public void visitNewClass(NewClassTree syntaxNode) {
       if (isOpeningResource(syntaxNode)) {
         final SymbolicValue instanceValue = programState.peekValue();
-        programState = programState.addConstraint(instanceValue, NullConstraint.OPENED);
+        if (instanceValue.wrappedValue() == instanceValue) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

File: java-squid/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java [hunk 4]
```diff
@@ -61,76 +59,55 @@
...
-    @Override
-    protected void scan(List<? extends Tree> trees) {
-      // Cut recursive processing
-    }
-
-    @Override
-    protected void scan(ListTree<? extends Tree> listTree) {
-      // Cut recursive processing
-    }
+  private static enum States {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java]: MAJOR Remove this redundant "static" qualifier. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2786)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 2 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Add a new line at the end of this file.](https://github.com/SonarSource/sonar-java/blob/9b17b5ee91b7d311f1a152b0012fb1f9fefa5acc/java-squid/src/main/java/org/sonar/java/se/checks/CheckerTreeNodeVisitor.java) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00113)

File: java-squid/src/main/java/org/sonar/java/se/ObjectConstraint.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    return new ObjectConstraint(isNull, syntaxNode, newState);
+  }
+
+  public boolean isNull() {
+    return isNull;
+  }
+
+  public boolean hasState(Object aState) {
+    if (aState == null) {
+      return state == null;
```
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/se/ObjectConstraint.java]: as state field is final and constructor parameter not annoted with @Nullable this should never be true.

File: java-squid/src/test/java/org/sonar/java/se/checks/UnclosedResourcesCheckTest.java~SONARJAVA-1315 - Unclosed resources [hunk 6]
```diff
@@ -1,46 +0,0 @@
-/*
```
Comment (benzonico) [inline java-squid/src/test/java/org/sonar/java/se/checks/UnclosedResourcesCheckTest.java~SONARJAVA-1315 - Unclosed resources]: This has been cleaned up on master... and as such should be removed from your commit.

**Comment by benzonico** [comment]
Merged in https://github.com/SonarSource/sonar-java/commit/633ff0d6f7b0d3141113282b46f8bad09445c99a
